build v9 directly

KurtYoung opened the pull request:
This PR tracks the feature of building v9 directly which had been discussed in https://groups.google.com/forum/#!topic/druid-development/0CxhljSGeeo

We can divide this PR into 3 main parts:

1. Changing of ColumnPartSerde's interface. I had change ColumnPartSerde's interface to separate the Serializer and Deserializer. So we can have multiple serializers, one of it has the same behavior with old ones. And another one is the new streaming version. The Deserializer just do the same thing like old ones.

2. Added IndexedIntsWriter interface and some *Writers which are responsible for writing dimension ids. You can see the hierarchy that i divided the writers to 2 main sub abstract classes: SingleValueIndexedIntsWriter and MultiValueIndexedIntsWriter. 
Here are the classed which are doing the real things:
VSizeIndexedIntsWriter (single value, vsize encoded, not compressed)
CompressedIntsIndexedWriter (single value, not vsize encoded, compressed)
CompressedVSizeIntsIndexedWriter (single value, vsize encoded, compressed)
VSizeIndexedWriter (multi value, both offset and values are vsized, not compressed)
CompressedVSizeIndexedV3Writer (multi value, only values are vsized, compressed)
More details can be found here: https://groups.google.com/forum/#!topic/druid-development/0CxhljSGeeo

3. GenericColumnSerializer and its sub-classes are for writing metrics. They are almost identical to MetricColumnSerializer you familiar with, except the GenericColumnSerializer can write to a Channel(which is allocated from smoosher when building). The sub-classes are:
LongColumnSerializer (write long metrics)
FloatColumnSerializer (write float metrics)
ComplexColumnSerializer (writer complex metrics)

nishantmonu51 commented on processing/src/main/java/io/druid/segment/data/CompressedIntsIndexedWriter.java:
```diff
@@ -0,0 +1,105 @@
+      flattener.write(StupidResourceHolder.create(endBuffer));
+    }
+    endBuffer = null;
+    flattener.close();
```
can we move close to a finally block.

fjy commented:
@KurtYoung This is an awesome PR and one that I know several folks want to code review. Everything is going to be slow during the Christmas holidays, but this will get attention after new years.

nishantmonu51 commented:
:+1:, seems good to go after a minor nitpick for closing in finally block. 
@KurtYoung thanks for the awesome contribution.

KurtYoung commented:
added IndexMergerV9, changed some low level interfaces but totally compatible with the old way.
*IndexMergerV9 has not been tested, just make the whole thing compiled and old tests passed* 

Some explanation and thoughts here: 
1. Changes to ColumnPartSerde's interface is to separate the Serializer and Deserializer. So we can have multiple serializers, one of it has the same behavior with old ones. And another one is the new streaming version.
2. All writers(time, dim, metrics) are backed by tmp files and copy to the final smoosh file directly.
3. GenericColumnSerializer and its sub-classes are almost identical to MetricColumnSerializer, except the GenericColumnSerializer can write to a Channel(which is allocated from smoosher when buding). The MetricColumnSerializer are used for build v8 index, and can be removed later.
4. IndexedIntsWriter are responsible for writing dimension ids which now has 4 types of the formats, more details are here: https://groups.google.com/forum/#!topic/druid-development/0CxhljSGeeo

And here are some points I think should be discussed with your guys when writing the codes:
1. about skippedDimensions in IndexMaker: dimensions with cardinality 0 are treated as skipped dimensions. But in current version of IncrementalIndex, even null dim value are converted to empty string and treated as a valid value, so I assume all the dimensions will be effective so there are no logic about skipped dimensions.
2. about nullSet when building inverted index both in IndexMaker and converting v8 to v9, it checks all dimension's null rows and make sure Dictionary and BitmapIndex are having these null rows. As mentioned in 1, the null values are treated properly in IncrementalIndex, so I did not handle this situation also.

binlijin commented on processing/src/main/java/io/druid/segment/data/CompressedIntsIndexedWriter.java:
```diff
@@ -0,0 +1,135 @@
+    if (!endBuffer.hasRemaining()) {
+      endBuffer.rewind();
+      flattener.write(StupidResourceHolder.create(endBuffer));
+      endBuffer.rewind();
```
New two IntBuffer to reuse, when after write switch them to make sure GenericIndexedWriter can sort correct.

binlijin commented on processing/src/main/java/io/druid/segment/IndexMergerV9.java:
```diff
@@ -0,0 +1,1285 @@
+      GenericIndexedWriter<String> writer = dimValueWriters.get(dimIndex);
+      int cardinality = 0;
+      for (String value : dimensionValues) {
+        value = value == null ? "" : value;
```
Not necessary.

KurtYoung commented:
hmm...I see the point why the null dict value and null row set are handled both in IndexMaker and IndexIO's converting. 

*My previous decision about skippedDimension & nullSet are wrong, just ignore it. *

Working on this now...

KurtYoung commented:
Found a bug of merge & maker about dimension orders:
https://github.com/druid-io/druid/issues/2162
I will add some unit tests and marked Ignore, let's just fix it in another PR

KurtYoung commented:
Is it ok to insert null to every dimension's dictionary even if the dimension did not contain any null values?
By doing this, there will be no need to iterate all rows just want to determine whether one dimension contains null value and to bump dictionary id if we did not.

navis commented:
Wishfully -1 could be used for null and number of distincts would be specified for each dimension in meta. Would be possible?

nishantmonu51 commented:
@KurtYoung: are there any corresponding changes in the filters/query path for null handling in case we add null to every dimension dictionary.

KurtYoung commented:
@nishantmonu51 I'm also aware of this, the current implementation did not add null to each dimension but handled null value in both IncrementalIndexAdapter and IndexMergerV9.

fjy commented:
@KurtYoung there's been a lot of optimizations in the old index merger over the last few months. Are those optimizations incorporated in building the v9 segment directly?

fjy commented on extensions/datasketches/src/main/java/io/druid/query/aggregation/datasketches/theta/SketchMergeComplexMetricSerde.java:
```diff
@@ -67,11 +65,10 @@ public Object extractValue(InputRow inputRow, String metricName)
 
   @Override
-  public ColumnPartSerde deserializeColumn(ByteBuffer buffer, ColumnBuilder builder)
+  public void deserializeColumn(ByteBuffer buffer, ColumnBuilder builder)
```
since we are changing the behavior of this method, can we please add a comment on the interface about how the method is supposed to be used?

fjy commented:
@KurtYoung to also help out with the review, can you update the description of the PR to include all the major changes you are making and their purpose?

fjy commented on processing/src/main/java/io/druid/segment/data/MultiValueIndexedIntsWriter.java:
```diff
@@ -0,0 +1,45 @@
+    } else if (obj instanceof List) {
+      addValues((List<Integer>) obj);
+    } else {
+      throw new IAE("unsupported object type");
```
can we print what the object type is as well?

fjy commented on processing/src/main/java/io/druid/segment/data/SingleValueIndexedIntsWriter.java:
```diff
@@ -0,0 +1,48 @@
+        addValue(vals[0]);
+      }
+    } else {
+      throw new IAE("unsupported object type");
```
same here, print the unsupported type

fjy commented on processing/src/main/java/io/druid/segment/data/TmpFileIOPeon.java:
```diff
@@ -34,8 +34,19 @@
 */
 public class TmpFileIOPeon implements IOPeon
 {
+  boolean allowConflict;
```
can we make this a private final? does allowConflict=true mean we can overwrite the tmp file? If so, can we call this isOverrideAllowed or something?

fjy replied:
you can use a Boolean and null check to set the default value

KurtYoung replied:
this option is useful when we merge index, if we used lots of temp files and did not want then to conflict.
The original behavior is allow us to overwrite the tmp file, but it's dangerous sometimes.

fjy commented on processing/src/main/java/io/druid/segment/serde/ComplexMetricSerde.java:
```diff
@@ -36,14 +36,11 @@
-   * added to the builder.
    *
    * @param buffer the buffer to deserialize
    * @param builder ColumnBuilder to add the column to
```
i think some more comments are needed here now that this interface has changed

KurtYoung replied:
Since the return value is not used anywhere, so i removed the return value.
IMO the first line of comments had explained the interface well:
"Deserializes a ByteBuffer and adds it to the ColumnBuilder."

fjy commented on processing/src/main/java/io/druid/segment/serde/DictionaryEncodedColumnPartSerde.java:
```diff
@@ -83,7 +87,169 @@ public byte asByte()
+    }
+  }
+
+  public static LegacySerializerBuilder legacySerializerBuilder()
```
instead of calling it legacy, can we assign the version to the serializer builder?

KurtYoung replied:
will do

KurtYoung replied:
any advises for naming the version? both serializer are suitable for v9 index, so we can not call the legacy V8 and the new one V9

fjy replied:
do we still need the old builder? I think we should switch to IndexMergerV9 as the default and remove the legacy stuff

fjy commented on processing/src/test/java/io/druid/segment/IndexMergerTest.java:
```diff
@@ -346,65 +406,6 @@ public void testPersistEmptyColumn() throws Exception
   }
 
   @Test
-  public void testMergeRetainsValues() throws Exception
```
why was this test removed?

KurtYoung replied:
my bad, revert it now

fjy replied:
the test is still removed :)

KurtYoung replied:
Its code place had been changed so it looks like this ut been removed, you can check out here:
https://github.com/druid-io/druid/pull/2138/files#diff-84cda5dd68d43d141ccedbbb7b3cbfd0

fjy commented on processing/src/test/java/io/druid/segment/IndexMergerTest.java:
```diff
@@ -749,6 +754,63 @@ public void testNonLexicographicDimOrderMerge() throws Exception
     Assert.assertArrayEquals(new Object[]{3L}, boatList.get(0).getMetrics());
   }
 
+  // test case for reproduce https://github.com/druid-io/druid/issues/2162
```
is this still an issue?

KurtYoung replied:
no, i will remove these test cases

fjy commented:
@KurtYoung I can't find the logic where you actually use index merger v9 instead of index merger

fjy commented:
@KurtYoung I think the best way to think about how to handle nulls/empty strings in Druid is described in this PR: https://github.com/druid-io/druid/pull/995

fjy commented:
I did a first pass over this PR but didn't go into detail for IndexMergerV9. Will look into more once we that know it reasonably works. High level I'm on board with the changes.

KurtYoung commented:
@fjy Actually, I did not change any logic to use IndexMergerV9 now, but I switch the current IndexMerger's logic to IndexMergerV9's and make all the test cases passes. 
I think maybe it's a good time to switch to IndexMergerV9 after you passed this PR.

fjy commented:
@KurtYoung can't seem to make any comments for indexmergerv9

fjy commented on processing/src/main/java/io/druid/segment/IndexMaker.java:
```diff
@@ -1365,7 +1374,10 @@ public Object apply(Rowboat input)
+                                  .build(),
             metBuilder,
             metric
         );
```
I am not able to make any comments for IndexMergerV9 below this line.

fjy replied:
is IndexMergerV9 just a rename of IndexMerger.java?

fjy replied:
If that's the case, did you remove the v8 to v9 conversion step in IndexMerger?

KurtYoung replied:
IndexMergerV9 make v9 index files directly, the main step are very like with IndexMerger except the v8 to v9 conversion step is no longer necessary.

fjy commented:
@KurtYoung indexmergerv9 seems to be indexMaker. Do you have any performance numbers of the improvements in indexing time? IndexMaker had many inefficiencies with how it would reiterate though the same data to build columns

KurtYoung commented:
@fjy I will give some performance number soon. And we had actually used IndexMergerV9 in our hadoop batch ingestion, the build time has decreased and memory more stable, no longer killed by yarn any more.

KurtYoung commented:
Here are some performance numbers which I tested on my laptop.
The input are 1 million rows with 100 dimension with high and low cardinality, and 20 simple metrics.
I set rowFlushBoundary to 100000 because of the memory pressure, so there will be 10 persist and 1 big merge.

The old way(create V8, then convert to V9):
total persist time: 101.3s ( 10 persists)
total merge time: 89.2s 

The new way(create V9 directly):
total persist time: 65.641113s
total merge time: 53.852s

The performance gain is about: (101.3+89.2-65.6-53.8)/(101.3+89.2) = 37.3%
The memory pressure is also reduced since we did not have to load all data into heap when converting v8 to v9.

fjy commented:
@KurtYoung These performance numbers are great. There's a few conflicts in the branch right now. Also, do you know why we can't comment in IndexMergerV9.java?

KurtYoung commented:
@fjy, i  think may be this is the reason: https://help.github.com/articles/what-are-the-limits-for-viewing-content-and-diffs-in-my-repository/
The limit is 1500 lines and current we have 1600+ lines. There are some static class like DimValueConverter and MMappedIndexRowIterable which are copied from IndexMerger, I can change these class to public static and use then from IndexMeger to reduce the diff line numbers.

KurtYoung commented:
Conflicts had been resolved, and we can make comments for IndexMergerV9 now.

fjy commented on processing/src/main/java/io/druid/segment/CompressedVSizeIndexedV3Supplier.java:
```diff
@@ -0,0 +1,138 @@
+ * If we want to streams VSizeInts, we must know the max value in the value sets. It's easy to know the max id of
+ * values(like dimension cardinality while encoding dimension), but difficult to known the max id of offsets.
+ */
+public class CompressedVSizeIndexedV3Supplier implements WritableSupplier<IndexedMultivalue<IndexedInts>>
```
if we replace IndexMerger with IndexMergerv9, we should also be able to replace CompressedVSizeIndexedSupplier with this no?

KurtYoung replied:
Are there any demands to read V8 index for some compatibility reason? if so, I think CompressedVSizeIndexedSupplier should still be used then.

xvrl replied:
Druid has retained backwards compatibility with all previous segments versions for seamless upgrade purposes. I believe we should maintain backwards compatibility here as well.

fjy commented on processing/src/main/java/io/druid/segment/IndexIO.java:
```diff
@@ -758,11 +757,21 @@ public int size()
               builder.setValueType(ValueType.LONG);
-              builder.addSerde(new LongGenericColumnPartSerde(holder.longType, BYTE_ORDER));
+              builder.addSerde(
+                  LongGenericColumnPartSerde.legacySerializerBuilder()
```
why can't we use the new builder here?

KurtYoung replied:
The main difference between the new and legacy builder is: The new builder holds some tmp file backed writer one should **add** every value to. Instead the legacy one directly holds the target data structure like CompressedLongsIndexedSupplier. In the convert case, we had read the structure out from the v8 index, so we just need to use the legacy builder to hold this structure and flush it out.

fjy commented on processing/src/main/java/io/druid/segment/IndexMergerV9.java:
```diff
@@ -0,0 +1,1321 @@
+import java.util.Set;
+import java.util.TreeSet;
+
+public class IndexMergerV9
```
I think this should replace the existing IndexMerger. Let's just have one version.

KurtYoung replied:
Agree

fjy commented on processing/src/main/java/io/druid/segment/IndexMergerV9.java:
```diff
@@ -0,0 +1,1321 @@
+    // We are materializing the list for performance reasons. Lists.transform
+    // only creates a "view" of the original list, meaning the function gets
+    // applied every time you access an element.
+    List<IndexableAdapter> indexAdapteres = Lists.newArrayList(
```
how much of a performance difference does this actually make? I think things are fine both ways though.

KurtYoung replied:
hmm. these comments are from https://github.com/druid-io/druid/pull/1960

fjy commented on processing/src/main/java/io/druid/segment/IndexMergerV9.java:
```diff
@@ -0,0 +1,1321 @@
+      throw new ISE("Couldn't make outdir[%s].", outDir);
+    }
+
+    // TODO: verify this
```
??

KurtYoung replied:
Oh, I forgot this, this TODO is left for issue https://github.com/druid-io/druid/issues/2162.
I noticed IndexMerger and IndexMaker treat the dimension order in different ways and I thought i should check this again.
So I think i will just do it in IndexMerger's way.

fjy commented on processing/src/main/java/io/druid/segment/IndexMergerV9.java:
```diff
@@ -0,0 +1,1321 @@
+        ++cardinality;
+      }
+
+      progress.progressSection(section, "dim[" + dimension + "] cardinality: " + cardinality);
```
very small nitpick, but can we use String.format instead to keep with the existing code guidelines

KurtYoung replied:
of cource

fjy commented on processing/src/main/java/io/druid/segment/IndexMergerV9.java:
```diff
@@ -0,0 +1,1321 @@
+    }
+    return Lists.newArrayList(retVal);
+  }
+}
```
This file reads more like IndexMaker in how it breaks up the indexing sections, which I like :)

I find this code much easier to read

fjy commented on processing/src/main/java/io/druid/segment/IndexMergerV9.java:
```diff
@@ -0,0 +1,1321 @@
+                 ? CompressedVSizeIntsIndexedWriter.create(ioPeon, filenameBase, cardinality, dimCompression)
+                 : new VSizeIndexedIntsWriter(ioPeon, filenameBase, cardinality);
+      }
+      writer.open();
```
can we add a comment here about the fact these are closed in another method later on? To make it easier to read for other folks working on this code later on

KurtYoung replied:
yes

KurtYoung replied:
added other places.

fjy commented on processing/src/main/java/io/druid/segment/IndexMergerV9.java:
```diff
@@ -0,0 +1,1321 @@
+          .addSerde(partBuilder.build())
+          .build();
+      makeColumn(v9Smoosher, dim, serdeficator);
+      progress.progressSection(section, String.format("Completed demension column[%s]", dim));
```
dimension* not demension

fjy commented on processing/src/main/java/io/druid/segment/IndexMergerV9.java:
```diff
@@ -0,0 +1,1321 @@
+    for (int i = 0; i < mergedDimensions.size(); ++i) {
+      nullRowsList.add(indexSpec.getBitmapSerdeFactory().getBitmapFactory().makeEmptyMutableBitmap());
+    }
+    walkThroughAndMergeRows(
```
technically this method actually writes the dimension and metric values

fjy replied:
how about we call this method mergeIndexesAndMakeColumns?

KurtYoung replied:
sounds good

fjy commented on processing/src/main/java/io/druid/segment/IndexMergerV9.java:
```diff
@@ -0,0 +1,1321 @@
+        nullRowsList, dimValueWriters, bitmapIndexWriters, spatialIndexWriters
+    );
+
+    /************ Build Columns *************/
```
Finalize building columns?

fjy commented on processing/src/main/java/io/druid/segment/IndexMergerV9.java:
```diff
@@ -0,0 +1,1321 @@
+        adapters, progress, mergedDimensions, dimCardinalities, dimValueWriters, dimensionSkipFlag, dimConversions
+    );
+
+    /************* Walk through data sets and merge them *************/
```
Walk through data sets, merge them, and write merged columns

fjy commented:
@KurtYoung At a high level I think this PR is great and I am :+1: after my comments are addressed. The biggest thing for me is that I think that we should only have one version of everything and we should just remove the legacy code.

xvrl commented:
Thanks @KurtYoung for working on this, this will be a very nice improvement.
I have just a few asks that might make reviewing this PR easier:
- Do you mind updating the PR description to reflect the overall changes you propose, so we don't have to read through all the comments?
- Would it be possible to break out separate PRs for the pieces that are independent, such as the addition of new compressedvsize classes? That will make reviewing the code change much easier and also help get things merged faster.
Thanks!

KurtYoung commented:
hi, @xvrl , we can divide this PR into 3 main parts:
1. Changing of ColumnPartSerde's interface. I had change ColumnPartSerde's interface to separate the Serializer and Deserializer. So we can have multiple serializers, one of it has the same behavior with old ones. And another one is the new streaming version. The Deserializer just do the same thing like old ones.

2. Added IndexedIntsWriter interface and some *Writers which are responsible for writing dimension ids. You can see the hierarchy that i divided the writers to 2 main sub abstract classes: SingleValueIndexedIntsWriter and MultiValueIndexedIntsWriter. 
Here are the classed which are doing the real things:
VSizeIndexedIntsWriter (single value, vsize encoded, not compressed)
CompressedIntsIndexedWriter (single value, not vsize encoded, compressed)
CompressedVSizeIntsIndexedWriter (single value, vsize encoded, compressed)

VSizeIndexedWriter (multi value, both offset and values are vsized, not compressed)
CompressedVSizeIndexedV3Writer (multi value, only values are vsized, compressed)

More details can be found here: https://groups.google.com/forum/#!topic/druid-development/0CxhljSGeeo

3.  GenericColumnSerializer and its sub-classes are for writing metrics. They are almost identical to MetricColumnSerializer you familiar with, except the GenericColumnSerializer can write to a Channel(which is allocated from smoosher when building). The sub-classes are:
LongColumnSerializer (write long metrics)
FloatColumnSerializer (write float metrics)
ComplexColumnSerializer (writer complex metrics)

KurtYoung commented:
@fjy All you comments had been taken care of, except i think maybe we can remove IndexMerger and some related classed in another PR. This a big enough PR to review, and changing like this will making another diff > 1500 lines so we can not comment again.

fjy commented:
@KurtYoung that makes sense, but I think it will be nice to submit another PR right after this to only have 1 version of things so we don't get into an indexmerger/indexmaker situation again. I know @cheddar also wants to review this PR as well.

KurtYoung commented:
@fjy Sure, I will fire another PR right after this one merged.

gianm commented:
IMO there is some value in having an option to toggle between the implementations for a while, as that would make it easier to test the new stuff and compare its output against the old stuff if necessary.

himanshug commented on indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java:
```diff
@@ -448,11 +448,12 @@ public void setConf(Configuration config)
     protected ProgressIndicator makeProgressIndicator(final Context context)
     {
-      return new BaseProgressIndicator()
+      return new LoggingProgressIndicator("IndexGeneratorReducer")
```
this will cause regression, pls see https://github.com/druid-io/druid/pull/1923 .

KurtYoung replied:
ok，going to revert this

himanshug commented on processing/src/main/java/io/druid/segment/data/IndexedIntsWriter.java:
```diff
@@ -0,0 +1,35 @@
+  public long getSerializedSize();
+
+  public void writeToChannel(WritableByteChannel channel) throws IOException;
+}
```
should this interface be called "ObjectsWriter" ? I don't see anything "indexed" or "int" specific in the interface.

himanshug replied:
ok, i see, name is more of a "intent marker" as the interface is pretty much same as GenericColumnSerializer.

KurtYoung replied:
yes, it's only responsible for writing dimension ids

himanshug commented on processing/src/main/java/io/druid/segment/data/VSizeIndexedIntsWriter.java:
```diff
@@ -0,0 +1,99 @@
+ */
+public class VSizeIndexedIntsWriter extends SingleValueIndexedIntsWriter
+{
+  private static final byte version = VSizeIndexedInts.version;
```
in general, static finals are conventionally upper case

himanshug replied:
this applies to other places as well.

KurtYoung replied:
ok

himanshug replied:
while going through the PR I realized that the convention wasn't followed in existing code and you probably copy/pasted most of it, so feel free to ignore this one.

KurtYoung replied:
Maybe i should change the existing code too.

himanshug commented on processing/src/main/java/io/druid/segment/data/VSizeIndexedIntsWriter.java:
```diff
@@ -0,0 +1,99 @@
+  )
+  {
+    this.ioPeon = ioPeon;
+    this.filenameBase = filenameBase;
```
can't we just store `filename = String.format("%s.%s", filenameBase, "values");` and not have makeFilename called in multiple places?

KurtYoung replied:
agree

himanshug commented on processing/src/main/java/io/druid/segment/data/CompressedIntsIndexedWriter.java:
```diff
@@ -0,0 +1,135 @@
+ */
+public class CompressedIntsIndexedWriter extends SingleValueIndexedIntsWriter
+{
+  public static final byte version = CompressedIntsIndexedSupplier.version;
```
private and VERSION

himanshug commented on processing/src/main/java/io/druid/segment/data/VSizeIndexedIntsWriter.java:
```diff
@@ -0,0 +1,99 @@
+  protected void addValue(int val) throws IOException
+  {
+    byte[] intAsBytes = Ints.toByteArray(val);
+    valuesOut.write(intAsBytes, intAsBytes.length - numBytes, numBytes);
```
since we use only last "numBytes" from intAsBytes, does it make sense to not create full 4 byte intAsBytes array?

KurtYoung replied:
It's easier to read and write this way, and i think one or two byte will not cause any heap pressure, so i kept it this way.

himanshug commented on processing/src/main/java/io/druid/segment/serde/ColumnPartSerde.java:
```diff
@@ -39,7 +39,19 @@
-  public long numBytes();
-  public void write(WritableByteChannel channel) throws IOException;
-  public ColumnPartSerde read(ByteBuffer buffer, ColumnBuilder builder, ColumnConfig columnConfig);
+  public interface Serializer
```
in general, druid code base keeps nested class/interface definitions in the end. it will be good to do same in here and other places.

himanshug commented on processing/src/main/java/io/druid/segment/serde/DictionaryEncodedColumnPartSerde.java:
```diff
@@ -83,7 +87,169 @@ public byte asByte()
 
-  public static class Builder
+  @JsonCreator
+  public static DictionaryEncodedColumnPartSerde createDeserializer(
```
now these methods in all impls should probably be called "createInstance" as "deserializer" is not what they return and that has different meaning after this PR.

himanshug commented on processing/src/main/java/io/druid/segment/serde/ComplexColumnPartSerde.java:
```diff
@@ -59,21 +58,110 @@ public String getTypeName()
 
-  @Override
-  public long numBytes()
+  public static SerializerBuilder serializerBuilder()
```
here and in other places, SerializerBuilder is really building [Complex]ColumnPartSerde and not "serializer" of the ColumnPartSerde. can we rename it to be ColumnPartSerdeBuilder instead?

KurtYoung replied:
It seems a little strange to me is the "Serde" name. What i thought before this is a abbreviation of serializer and deserializer, does it?
So i think ColumnPartSerde is a combination of Serializer and Deserializer, when we create object, it's either a Serializer or Deserializer

himanshug replied:
I think my confusion mostly arose because ColumnPartSerde now declares new interfaces called Serializer/Deserializer and SerializerBuilder does not build "Serializer" instance but "ComplexMetricSerde" instance. I guess, its a matter of perspective.
you can leave it as is then.

himanshug commented on processing/src/main/java/io/druid/segment/IndexMergerV9.java:
```diff
@@ -0,0 +1,1277 @@
+
+    progress.progress();
+    final Map<String, ValueType> metricsValueTypes = Maps.newTreeMap(Ordering.<String>natural().nullsFirst());
+    final Map<String, String> metricTypeNames = Maps.newTreeMap(Ordering.<String>natural().nullsFirst());
```
can we call this complexMetricTypeNames ?

KurtYoung replied:
it's a type name map for all types of the metrics, not only for the complex ones.

himanshug replied:
ok, yeah, but it is only used for complex metrics. for others, value type is specific enough.

himanshug commented on processing/src/main/java/io/druid/segment/IndexMergerV9.java:
```diff
@@ -0,0 +1,1277 @@
+    /************* Setup Dim Conversions **************/
+    progress.progress();
+    final Map<String, Integer> dimCardinalities = Maps.newHashMap();
+    final ArrayList<GenericIndexedWriter<String>> dimValueWriters = setupDimValueWriters(ioPeon, mergedDimensions);
```
can we replace "dimValue" by "dictionary" ?

KurtYoung replied:
IMO "dimValue" is more consistent with other classes who also treat these as dim values other than dictionary

himanshug replied:
ok, agree with consistency.

himanshug commented:
aside from some minor comments, this PR looks good to me. However, I would agree with @gianm about not completely replacing IndexMerger with this but release them together and give user ability to choose. After a while, when it has been used in production in multiple clusters, we can remove IndexMerger.
One quick way of doing that would be to make IndexMergerV9 extend IndexMerger with all the relevant methods overridden, then based on a runtime.properties configuration bind either IndexMerger or IndexMergerV9 instance in guice.

Side-note: I hadn't really looked at IndexMaker code much and curious to know at high level what high level improvements make this better than IndexMaker? can anyone with relevant knowledge pls answer.

fjy commented:
@himanshug IndexMaker iterates through every row each time it tries to build a column. This is a much more efficient implementation.

KurtYoung commented:
@himanshug Another reason is IndexMaker keep all needed values in memory when it tries to build a column, it will cause memory pressure and hurt performance.

fjy commented:
@himanshug Any more comments?

@KurtYoung Can we add some way to switch between IndexMerger and IndexMerger v9 in the configuration? Index Merger should be the default.

himanshug commented:
@fjy I am :+1: once the configuration to switch to IndexMergerV9 is in place.

xvrl commented:
I believe @cheddar should have a look at this one. He had a lot of opinions about format when I made changes to introduce dimension compression, and this one introduces even more changes.

I also agree with @gianm we should be able to switch between implementations until it has been verified to be production ready.

xvrl commented:
I believe @cheddar should have a look at this one. He had a lot of opinions about format when I made changes to introduce dimension compression, and this one introduces even more changes.

I also agree with @gianm we should be able to switch between implementations until it has been verified to be production ready.

KurtYoung commented:
@fjy ok, i will add a config about this. Also looking forward to @cheddar 's comments :)

KurtYoung commented:
Added "buildV9Directly" config to TuningConfig, docs are updated

fjy commented:
@himanshug @xvrl we good to move forward?

himanshug commented:
:+1: for me

fjy commented:
:+1: for me too. Will leave this open until tomm to see if anyone else has comments.

@KurtYoung have you filled out the CLA: http://druid.io/community/cla.html

You guys might consider a corporate CLA.

himanshug commented:
@fjy @KurtYoung pls squash the commits / cleanup the history, very useful contribution.

xvrl commented:
@fjy @KurtYoung I might be wrong, but I still see some comments outstanding. Can we respond or address them?
- https://github.com/druid-io/druid/pull/2138/files#r48401969
- https://github.com/druid-io/druid/pull/2138#discussion_r48784768
- https://github.com/druid-io/druid/pull/2138#discussion_r48803287
- https://github.com/druid-io/druid/pull/2138#discussion_r49127146
- https://github.com/druid-io/druid/pull/2138#discussion_r49095885

KurtYoung commented:
@fjy I have filled out individual CLA, don't know if i had the right to fill a corporate CLA.

fjy commented:
@KurtYoung thanks

@xvrl any more comments?

KurtYoung commented:
@xvrl All these comments had been addressed and solved.

KurtYoung commented:
squashed into 3 commits.

xvrl commented on indexing-hadoop/src/main/java/io/druid/indexer/HadoopTuningConfig.java:
```diff
@@ -191,6 +196,11 @@ public boolean getUseCombiner()
   }
 
+  @JsonProperty
+  public Boolean getBuildV9Directly() {
```
since buildV9Directly is never null, this should probably be `boolean` instead of `Boolean`

xvrl replied:
and also renamed to `isBuildV9Directly`

xvrl commented on indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerConfig.java:
```diff
@@ -351,6 +354,11 @@ public HadoopyShardSpec getShardSpec(Bucket bucket)
 
+  public boolean isBuildV9Directly()
+  {
+    return schema.getTuningConfig().getBuildV9Directly();
```
this will throw an NPE if getBuildV9Directly returns null

xvrl commented on indexing-service/src/main/java/io/druid/indexing/common/task/RealtimeIndexTask.java:
```diff
@@ -286,6 +287,9 @@ public String getVersion(final Interval interval)
     );
     this.queryRunnerFactoryConglomerate = toolbox.getQueryRunnerFactoryConglomerate();
 
+    IndexMerger indexMerger = spec.getTuningConfig().getBuildV9Directly()
```
this will throw an NPE if getBuildV9Directly returns null.

xvrl replied:
instead I think we should make the tuning config use `boolean`

xvrl commented on indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java:
```diff
@@ -353,12 +355,15 @@ public DataSegment push(File file, DataSegment segment) throws IOException
     final FireDepartmentMetrics metrics = new FireDepartmentMetrics();
     final Firehose firehose = firehoseFactory.connect(ingestionSchema.getDataSchema().getParser());
     final Supplier<Committer> committerSupplier = Committers.supplierFromFirehose(firehose);
+    final IndexMerger indexMerger = ingestionSchema.getTuningConfig().getBuildV9Directly()
```
this will throw an NPE if getBuildV9Directly returns null
